src/webots/nodes/utils/WbNodeUtilities.cpp

```diff
@@ -663,7 +663,24 @@ WbMatter *WbNodeUtilities::findUppermostMatter(WbNode *node) {
 }
 
 WbSolid *WbNodeUtilities::findTopSolid(const WbNode *node) {
-  return dynamic_cast<WbSolid *>(const_cast<WbNode *>(findTopNode(node)));
+  if (node == NULL || node->isWorldRoot())
+    return NULL;
+
+  const WbNode *n = node;
+  const WbNode *parent = n->parentNode();
+  WbSolid *topSolid = NULL;
+  WbSolid *currentNode;
+  while (parent) {
+    currentNode = dynamic_cast<WbSolid *>(const_cast<WbNode *>(n));
+    if (currentNode)
+      topSolid = currentNode;
+    if (parent->isWorldRoot())
+      return topSolid;
+
+    n = parent;
+    parent = n->parentNode();
+  }
+  return NULL;
```

Comment:
I don't know if you must return `NULL` or `topSolid`. Could the code have found a top Solid without having meet the WorldRoot ?
Moreover, if you return `topSolid`, you could make a `break` instead of the `return` on line 678, just to get only one exit point in the function, which I found to be kind of nicer :)